llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -3390,7 +3423,8 @@ class BoUpSLP {
     Last->State = EntryState;
     // FIXME: Remove once support for ReuseShuffleIndices has been implemented
     // for non-power-of-two vectors.
-    assert((has_single_bit(VL.size()) || ReuseShuffleIndices.empty()) &&
+    assert((hasFullVectorsOnly(*TTI, getValueType(VL.front()), VL.size()) ||
```

Comment:
Have you audited the reordering code for VL=N*VLMAX?  I suspect not as it's not trivial, and I'd expect it to require changes.  If you have, we should be able to just remove this assert instead as it should work for any VL.

Reply:
This is not reordering, this is reusing of the same scalars. The test checks it (see `[[TMP1:%.*]] = shufflevector <6 x i64> [[TMP0]], <6 x i64> poison, <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 0, i32 0>`)